node/src/test/scala/coop/rchain/node/hardening/Hardening.scala

```diff
@@ -0,0 +1,524 @@
+    rm  <- Resource.liftF(Resources.mkRuntimeManagerAt[Task](kvm))
+  } yield rm
+
+  type BalanceSheet = Map[User, (Long, Seq[Payment])]
+
+  val validatorUsers = validatorKeyPairs.map {
+    case (_, pk) => User(PrivateKey(ByteString.EMPTY), pk, Base16.encode(pk.bytes))
+  }
+
+  def verifyBalances[F[_]: Concurrent](balances: BalanceSheet, state: StateHash)(
+      implicit runtimeManager: RuntimeManager[F]
+  ): F[Unit] = {
+    def getVault(addr: String): String =
+      s"""new return, rl(`rho:registry:lookup`), RevVaultCh, vaultCh, balanceCh in {
+         |  rl!(`rho:rchain:revVault`, *RevVaultCh) |
+         |  for (@(_, RevVault) <- RevVaultCh) {
+         |    @RevVault!("findOrCreate", "${addr}", *vaultCh) |
+         |    for (@(true, vault) <- vaultCh) {
+         |      return!(vault)
+         |    }
+         |  }
+         |}
+         |""".stripMargin
+
+    Stream
+      .fromIterator(balances.iterator)
+      .filterNot {
+        // do not check per validator vaults
+        case (User(_, _, addr), _) => validatorUsers.map(_.addr).contains(addr)
+      }
+      .parEvalMapProcBounded {
+        case (User(_, _, addr), (paperBalance, txs)) =>
+          for {
+            vaultPar    <- runtimeManager.playExploratoryDeploy(getVault(addr), state)
+            runtime     <- runtimeManager.spawnRuntime
+            _           <- runtime.reset(Blake2b256Hash.fromByteString(state))
+            realBalance <- VaultBalanceGetter.getBalanceFromVaultPar(vaultPar.head, runtime)
+            _ = if (realBalance.get != paperBalance) {
+              println(s"TXs for ${addr}:")
+              println(s"${txs.mkString("\n")}")
+            }
+            _ = realBalance.get shouldBe paperBalance
+          } yield ()
+      }
+      .compile
+      .lastOrError
+  }
+
+  "merging with leader" should "work" in effectTest {
+
+    runtimeManagerResource.use { runtimeManager =>
+      {
+        implicit val c                                 = Concurrent[Task]
+        implicit val rm: RuntimeManager[Task]          = runtimeManager
+        implicit val metrics: Metrics.MetricsNOP[Task] = new Metrics.MetricsNOP
+        val kvm                                        = new InMemoryStoreManager
+
+        val mergesNum = 1
+        for {
+          dagStore <- BlockDagKeyValueStorage.create[Task](kvm)
+          _        <- dagStore.insert(genesis, false, true)
+
+          initUserBalances = genesisContext.genesisVaults.map {
+            case (_, pubKey) =>
+              (
+                users.find(_.pk == pubKey).get,
+                (GenesisBuilder.predefinedVaultsAmt, Seq.empty[Payment])
+              )
+          }.toMap
+          initValidatorBalances = validatorUsers.map(_ -> (0L, List.empty)).toMap
+          initBalances          = initUserBalances ++ initValidatorBalances
+
+          _ <- (genesis, 0, initBalances).tailRecM[Task, (BlockMessage, BalanceSheet)] {
```

Comment:
This is the main recursion that advances the DAF, one layer after another. Code inside is quite chaotic, but the intention is slowly move towards tagless final description (WIP is in comments at the top of the file) allowing building arbitrary DAGs, so for now I suggest leave it as it is.